### PR TITLE
GTesting is finally here

### PR DESCRIPTION
**Pull request by kamilritz**
This PR is adding the possibility to test the ECL with Google GTest library. 
A first basic test of the EKF is implemented to check if the filter is initializing properly under default sensor input.

@jkflying @dagar @bresch

### CODE REVIEW

**Comment by dagar** [comment]
Nice!

**Comment by dagar** [comment]
Something related, the python swig testing has been a bit of a maintenance headache. We should review what it's actually doing and consider replacing or dropping it.

**Comment by kamilritz** [comment]
The CMakeLists.txt's are not yet setup correctly. The GTests are also run with plain `make` and this causes the build on OSX to fail.

**Comment by dagar** [comment]
> The CMakeLists.txt's are not yet setup correctly. The GTests are also run with plain `make` and this causes the build on OSX to fail.

For simplicity I would try and mirror the general gtest setup from PX4/Firmware. https://github.com/PX4/Firmware/tree/master/cmake/gtest

File: test/CMakeLists.txt.in [hunk 1]
```diff
@@ -0,0 +1,19 @@
+cmake_minimum_required(VERSION 2.8.4)
```
Comment (TSC21) [inline test/CMakeLists.txt.in]: Are we making this mandatory? What if one already has Gtest on its system?

Comment (jkflying) [reply test/CMakeLists.txt.in]: This is the same way that the Firmware does it, it means you aren't dependent on system setup as much.

**Comment by kamilritz** [comment]
It is now in a state in which it could get in.
@dagar @jkflying @bresch @TSC21 Please review this, so that we can get this in soon and also other people can start to add tests!

File: CMakeLists.txt [hunk 2]
```diff
@@ -90,20 +90,19 @@ if(CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR)
 	elseif(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX)
 		add_compile_options(
 			-pedantic
-		
+
 			-Wall
-			-Wextra
+			#-Wextra
```
Comment (jkflying) [inline CMakeLists.txt]: Add a comment why this was disabled

File: Makefile [hunk 3]
```diff
@@ -98,19 +98,19 @@ doxygen:
...
 test_EKF: test_build
 	@cmake --build $(SRC_DIR)/build/test_build --target ecl_EKF_pytest-quick
-	
+
 test_EKF_plots: test_build
 	@cmake --build $(SRC_DIR)/build/test_build --target ecl_EKF_pytest-plots
 
 test_build_asan:
-	@$(call cmake-build,$@,$(SRC_DIR), "-DECL_ASAN=ON")
+	@$(call cmake-build,$@,$(SRC_DIR), "-DECL_ASAN=ON -DBUILD_TESTING=ON")
```
Comment (jkflying) [inline Makefile]:
```suggestion
	@$(call cmake-build,$@,$(SRC_DIR), "-DECL_ASAN=ON", "-DBUILD_TESTING=ON")
```

File: test/main.cpp [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#include <gtest/gtest.h>
+
+int main(int argc, char **argv) {
+  testing::InitGoogleTest(&argc, argv);
+  std::cout << "Run ECL gtests \n";
```
Comment (jkflying) [inline test/main.cpp]:
```suggestion
  std::cout << "Run ECL gtests" << std::endl;
```

**Comment by dagar** [comment]
Code coverage: https://codecov.io/gh/PX4/ecl/compare/e09e3e17a14f210a6e3c1f988336fede0d5b377f...1ccf8d12cd3fecd9dc3c48a72a1bd211269a23c3/changes

File: CMakeLists.txt (hunk 2 above)
Comment (kamilritz) [reply CMakeLists.txt]: Done

File: Makefile (hunk 3 above)
Comment (kamilritz) [reply Makefile]: Done

**Comment by jkflying** [comment]
@dagar Jenkins should be using the Jenkinsfile in the branch, right? It's still trying to point to the old Pythontest target.

**Comment by dagar** [comment]
> @dagar Jenkins should be using the Jenkinsfile in the branch, right? It's still trying to point to the old Pythontest target.

Not necessarily, for security it uses the non-modified Jenkinsfile for outside contributors. I'll get @kamilritz added to the org.

**Comment by bresch** [comment]
@kamilritz awesome, I'm looking forward to adding more unit tests!

**Comment by kamilritz** [comment]
@dagar Thanks a lot, CI is passing now. LGTM

**Pull request merged**